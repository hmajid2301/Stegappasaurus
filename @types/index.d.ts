import { ImageSourcePropType } from "react-native";

export type AlgorithmNames = "LSB";
export type BackgroundColors = "000" | "#FFF";
export type ThemeColors = "#17212D" | "#FFF";
export type PrimaryColor = "#009CFF" | "#E88C0C";
export type PrimaryColorNames = "ORANGE" | "BLUE";
export type PossibleAppStates = "inactive" | "background" | "active";

export interface ITheme {
  background: BackgroundColors;
  color: ThemeColors;
  isDark: boolean;
}

export interface IIcon {
  color: string;
  name: string;
  size?: number;
  type:
    | "material"
    | "material-community"
    | "font-awesome"
    | "octicon"
    | "ionicon"
    | "foundation"
    | "evilicon"
    | "simple-line-icon"
    | "zocial"
    | "entypo"
    | "feather"
    | "antdesign";
}

interface ISlide {
  color: string;
  height?: number;
  image: ImageSourcePropType;
  key: string;
  text: string;
  title: string;
  width?: number;
}
