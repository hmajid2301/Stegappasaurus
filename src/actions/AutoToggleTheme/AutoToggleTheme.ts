import AsyncStorage from "@react-native-community/async-storage";
import * as Location from "expo-location";
import * as Permissions from "expo-permissions";
import { getSunrise, getSunset } from "sunrise-sunset-js";

import Snackbar from "~/actions/Snackbar";

export default class AutoToggleTheme {
  private static oneDay = 24 * 60 * 60 * 1000;

  public async shouldToggleDarkTheme() {
    const currentTime = new Date(Date.now());
    const { sunrise, sunset } = await this.getSunriseAndSunsetTime(currentTime);
    let toggleTheme = true;

    if (sunrise !== null && sunset !== null) {
      if (currentTime > sunrise && currentTime < sunset) {
        toggleTheme = false;
      }
    }

    return toggleTheme;
  }

  private async getSunriseAndSunsetTime(currentTime: Date) {
    const { latitude, longitude } = await this.getLatitudeLongitude();
    let sunset = null;
    let sunrise = null;

    if (!isNaN(latitude)) {
      sunset = getSunset(latitude, longitude, currentTime);
      sunrise = getSunrise(latitude, longitude, currentTime);

      if (sunrise > sunset) {
        sunset = new Date(sunset.getTime() + AutoToggleTheme.oneDay);
      }
    }
    return { sunset, sunrise };
  }

  private async getLatitudeLongitude() {
    const currentDate = new Date(Date.now());
    const lastQueried = await AsyncStorage.getItem("@LastQueriedLocation");
    let latitude: number;
    let longitude: number;
    let lastQueriedDate: Date;

    if (lastQueried) {
      lastQueriedDate = new Date(lastQueried);
    } else {
      lastQueriedDate = new Date(
        currentDate.getTime() - AutoToggleTheme.oneDay
      );
    }

    if (
      currentDate.getTime() - lastQueriedDate.getTime() >=
      AutoToggleTheme.oneDay
    ) {
      ({ latitude, longitude } = await this.getNewLatitudeLongitude(
        currentDate
      ));
    } else {
      latitude = Number((await AsyncStorage.getItem("@Latitude")) || "51.5");
      longitude = Number((await AsyncStorage.getItem("@Longitude")) || "-0.12");
    }

    return { latitude, longitude };
  }

  private async getNewLatitudeLongitude(currentDate: Date) {
    let latitude;
    let longitude;
    const { status } = await Permissions.askAsync(Permissions.LOCATION);
    if (status === "granted") {
      const locationOn = await Location.hasServicesEnabledAsync();

      if (locationOn) {
        const location = await Location.getCurrentPositionAsync({});
        latitude = location.coords.latitude;
        longitude = location.coords.longitude;

        await Promise.all([
          AsyncStorage.setItem("@Latitude", JSON.stringify(latitude)),
          AsyncStorage.setItem("@Longitude", JSON.stringify(longitude)),
          AsyncStorage.setItem(
            "@LastQueriedLocation",
            JSON.stringify(currentDate)
          )
        ]);
      }
    }

    if (latitude === undefined || longitude === undefined) {
      throw Error("No location found");
    }

    return { latitude, longitude };
  }
}
