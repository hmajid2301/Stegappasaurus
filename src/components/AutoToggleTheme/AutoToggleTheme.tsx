import * as Location from "expo-location";
import * as Permissions from "expo-permissions";
import moment from "moment";
import { AsyncStorage } from "react-native";
import { getSunrise, getSunset } from "sunrise-sunset-js";

import Snackbar from "~/components/Snackbar";

export default class AutoToggleTheme {
  public shouldToggleTheme = async () => {
    const { sunrise, sunset } = await this.getSunriseAndSunsetTime();
    let toggleTheme = false;

    if (sunrise !== null && sunset !== null) {
      const currentTime = new Date();

      if (currentTime > sunrise && currentTime < sunset) {
        toggleTheme = true;
      }
    }

    return toggleTheme;
  };

  private getSunriseAndSunsetTime = async () => {
    const { latitude, longitude } = await this.getLatitudeLongitude();

    if (isNaN(latitude)) {
      Snackbar.show({
        text:
          "To use the automatic theme feature, location services must be turned on. Turning off automatic theme."
      });
      return { sunset: null, sunrise: null };
    }

    const sunset = getSunset(latitude, longitude);
    const sunrise = getSunrise(latitude, longitude);
    return { sunset, sunrise };
  };

  private async getLatitudeLongitude() {
    const currentDate = moment().toISOString();
    const lastQueried = await AsyncStorage.getItem("@LastQueriedLocation");
    let lastQueriedDate = currentDate;

    if (lastQueried !== null) {
      lastQueriedDate = moment(lastQueried)
        .add(7, "days")
        .toISOString();
    }

    let latitude = parseInt(
      (await AsyncStorage.getItem("@Latitude")) || "0",
      10
    );
    let longitude = parseInt(
      (await AsyncStorage.getItem("@Longitude")) || "0",
      10
    );

    if (lastQueriedDate > currentDate) {
      ({ latitude, longitude } = await this.getNewLatitudeLongitude(
        currentDate,
        latitude,
        longitude
      ));
    }
    return { latitude, longitude };
  }

  private async getNewLatitudeLongitude(
    currentDate: string,
    latitude: number,
    longitude: number
  ) {
    const { status } = await Permissions.askAsync(Permissions.LOCATION);
    if (status === "granted") {
      const locationOn = await Location.hasServicesEnabledAsync();
      if (locationOn) {
        const location = await Location.getCurrentPositionAsync({
          accuracy: Location.Accuracy.Lowest
        });

        latitude = location.coords.latitude;
        longitude = location.coords.longitude;
        await AsyncStorage.setItem("@Latitude", JSON.stringify(latitude));
        await AsyncStorage.setItem("@Longitude", JSON.stringify(longitude));
        await AsyncStorage.setItem("@LastQueriedLocation", currentDate);
      }
    }
    return { latitude, longitude };
  }
}
