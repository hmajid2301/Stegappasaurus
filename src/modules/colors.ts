export const colors = {
  darkColor: "#17212D",
  iconBlack: "#222",
  iconBlue: "#007CFE",
  iconGreen: "#4CDA64",
  iconLightGreen: "#25AE88",
  iconOrange: "#FF9501",
  iconRed: "#FD3C2D",
  primary: "#009CFF",
  pureBlack: "#000",
  pureWhite: "#FFF",
  secondary: "#E88C0C"
};
