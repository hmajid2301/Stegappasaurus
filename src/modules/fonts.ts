export const fonts = {
  body: "Roboto",
  bodyLight: "Roboto-Light",
  header: "Roboto-Thin"
};
