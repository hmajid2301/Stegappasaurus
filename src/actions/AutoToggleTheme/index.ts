import AutoToggleTheme from "./AutoToggleTheme";

export default AutoToggleTheme;
