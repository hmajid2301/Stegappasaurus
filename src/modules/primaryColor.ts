import { PrimaryColor, PrimaryColorNames } from "@types";
import { colors } from "./colors";

interface IPrimaryColor {
  name: PrimaryColorNames;
  color: PrimaryColor;
}

export const PRIMARY_COLOR: IPrimaryColor = {
  color: colors.primary as PrimaryColor,
  name: "BLUE"
};
export const SECONDARY_COLOR: IPrimaryColor = {
  color: colors.secondary as PrimaryColor,
  name: "ORANGE"
};
