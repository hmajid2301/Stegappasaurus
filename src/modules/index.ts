import { colors } from "./colors";
import { fonts } from "./fonts";
import { PRIMARY_COLOR, SECONDARY_COLOR } from "./primaryColor";
import { DARK_THEME, PRIMARY_THEME } from "./themes";

export {
  colors,
  PRIMARY_COLOR,
  SECONDARY_COLOR,
  PRIMARY_THEME,
  DARK_THEME,
  fonts
};
