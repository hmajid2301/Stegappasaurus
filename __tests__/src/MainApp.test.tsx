import AsyncStorage from "@react-native-community/async-storage";
import { mount, shallow, ReactWrapper } from "enzyme";
import React from "react";
import configureStore, { MockStoreEnhanced } from "redux-mock-store";

import { ThemeColors } from "@types";
import AutoToggleTheme from "~/actions/AutoToggleTheme";
import { colors } from "~/modules";
import MainAppConnected, { MainApp } from "~/MainApp";

const toggleDarkTheme = (isDark: boolean) => {
  isDark = !isDark;
};

const toggleAutomaticTheme = (isAutomatic: boolean) => {
  isAutomatic = !isAutomatic;
};

const props = {
  isAutomatic: true,
  toggleAutomaticTheme: (isAutomatic: boolean) => toggleAutomaticTheme(true),
  toggleDarkTheme: (isDark: boolean) => toggleDarkTheme(true),
  theme: {
    background: colors.darkColor as ThemeColors,
    color: colors.pureWhite as ThemeColors,
    isDark: true
  }
};

jest.mock("~/views/Routes", () => "App");
describe("MainApp: Match snapshots", () => {
  test("1: Normal", () => {
    const component = shallow(<MainApp {...props} />);
    expect(component).toMatchSnapshot();
  });

  test("2: Loading", () => {
    const component = shallow(<MainApp {...props} />);
    component.setState({ loading: true });
    expect(component).toMatchSnapshot();
  });

  test("3: No Intro Slides", () => {
    const component = shallow(<MainApp {...props} />);
    component.setState({ introShown: "true" });
    expect(component).toMatchSnapshot();
  });
});

describe("MainApp: Functions", () => {
  let component: ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;
  let instance: React.Component<{}, {}, any>;

  beforeAll(async () => {
    component = mount(<MainApp {...props} />);
    instance = component.instance();
  });

  const testData = [
    {
      introShown: null,
      expectedResult: false
    },
    {
      introShown: "false",
      expectedResult: false
    },
    {
      introShown: "true",
      expectedResult: true
    }
  ];
  test.each(testData)(
    "componentDidMount",
    async ({ expectedResult, introShown }) => {
      if (introShown) {
        await AsyncStorage.setItem("@IntroShown", introShown);
      }
      const componentSpy = await jest.spyOn(
        MainApp.prototype,
        "componentDidMount"
      );

      await (instance as any).componentDidMount();
      expect(component.state("loading")).toBe(false);
      expect(component.state("introShown")).toBe(expectedResult);
      expect(componentSpy).toHaveBeenCalled();
    }
  );

  test("componentWillUnmount", () => {
    const componentSpy = jest.spyOn(MainApp.prototype, "componentWillUnmount");
    (instance as any).componentWillUnmount();
    expect(componentSpy).toHaveBeenCalled();
  });

  test("introShownToUser", async () => {
    await (instance as any).introShownToUser();
    expect(component.state("introShown")).toBe(true);
  });

  const appData = [
    {
      appInFocus: "active",
      timesCalled: 1
    },
    {
      appInFocus: "background",
      timesCalled: 0
    }
  ];
  test.each(appData)("appInFocus", async ({ appInFocus, timesCalled }) => {
    const toggleSpy = jest
      .spyOn(AutoToggleTheme.prototype, "shouldToggleDarkTheme")
      .mockResolvedValue(true);
    await (instance as any).appInFocus(appInFocus);

    expect(toggleSpy).toHaveBeenCalledTimes(timesCalled);
    toggleSpy.mockClear();
  });
});

describe("MainApp Redux", () => {
  let store: MockStoreEnhanced<unknown, {}>;
  beforeAll(() => {
    const mockStore = configureStore();
    const initialState = {
      ToggleDarkTheme: {
        theme: props["theme"]
      },
      ToggleAutomaticTheme: {
        isAutomatic: true
      }
    };
    store = mockStore(initialState);
  });

  test("mapStateToProps", async () => {
    const component = shallow(
      //@ts-ignore
      <MainAppConnected store={store} />
    ).dive();
    expect((component.props() as any).theme).toEqual(props["theme"]);
    expect((component.props() as any).isAutomatic).toEqual(true);
  });

  test("mapDispatchToProps", async () => {
    const component = shallow(
      //@ts-ignore
      <MainAppConnected store={store} />
    ).dive();
    (component.props() as any).toggleDarkTheme(true);
    const actions = store.getActions();
    expect(actions).toMatchSnapshot();
  });
});
