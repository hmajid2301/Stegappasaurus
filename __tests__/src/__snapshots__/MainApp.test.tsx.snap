// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainApp Redux mapDispatchToProps 1`] = `
Array [
  Object {
    "payload": Object {
      "isDark": true,
    },
    "type": "TOGGLE_DARK_THEME",
  },
]
`;

exports[`MainApp: Match snapshots 1: Normal 1`] = `
<StatusBar
  animated={false}
  hidden={true}
  showHideTransition="fade"
>
  <Loader
    loading={true}
  />
</StatusBar>
`;

exports[`MainApp: Match snapshots 2: Loading 1`] = `
<StatusBar
  animated={false}
  hidden={true}
  showHideTransition="fade"
>
  <Loader
    loading={true}
  />
</StatusBar>
`;

exports[`MainApp: Match snapshots 3: No Intro Slides 1`] = `
<StatusBar
  animated={false}
  hidden={true}
  showHideTransition="fade"
>
  <Loader
    loading={true}
  />
</StatusBar>
`;
